compiler/rustc_mir_build/src/build/matches/mod.rs

```diff
@@ -1811,13 +1800,12 @@ impl<'a, 'tcx> Builder<'a, 'tcx> {
         let otherwise_post_guard_block = otherwise_candidate.pre_binding_block.unwrap();
         self.break_for_else(otherwise_post_guard_block, else_target, self.source_info(expr_span));
 
-        self.declare_bindings(
-            source_scope,
-            pat.span.to(span),
-            pat,
-            ArmHasGuard(false),
-            opt_expr_place,
-        );
+        let scope =
+            self.declare_bindings(None, pat.span.to(span), pat, ArmHasGuard(false), opt_expr_place);
+
+        if let Some(scope) = scope {
+            self.source_scope = scope;
+        }
```

Comment:
This is done in `lower_match` for `match` statements, so I do the same here.

Reply:
actually, this seems insufficient to solve the underlying issue, though its close.